three-cards-divination/src/main/java/com/cyber/mysticism/tarot/controller/ThreeCardsDivinationController.java

```diff
@@ -1,59 +1,55 @@
+import io.swagger.v3.oas.annotations.media.Content;
+import io.swagger.v3.oas.annotations.media.Schema;
+import io.swagger.v3.oas.annotations.responses.ApiResponse;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.RequestHeader;
 import org.springframework.web.bind.annotation.RestController;
 
-import java.nio.charset.StandardCharsets;
-import java.util.Base64;
-import java.util.Map;
-
 @RestController
 public class ThreeCardsDivinationController {
 
     public static final String UNAUTHORIZED = "Unauthorized!";
     private final ThreeCardsDivinationService threeCardsDivinationService;
+    private final AuthenticationUtil authenticationUtil;
 
     @Autowired
-    public ThreeCardsDivinationController(ThreeCardsDivinationService threeCardsDivinationService) {
+    public ThreeCardsDivinationController(ThreeCardsDivinationService threeCardsDivinationService,
+                                          AuthenticationUtil authenticationUtil) {
         this.threeCardsDivinationService = threeCardsDivinationService;
+        this.authenticationUtil = authenticationUtil;
     }
 
-    @GetMapping("/three-cards-divination")
-    public Map<String, Card> performReadingForUser(@RequestHeader(value = "Authorization", required = false) String authHeader)
+    @Operation(summary = "Perform a reading for an user", description = "Perform a three cards divination reading for an authenticated user using basic auth")
+    @ApiResponse(responseCode = "200", description = "Reading performed successfully",
+            content = @Content(mediaType = "application/json", schema = @Schema(implementation = ThreeCardsDivinationReading.class)))
+    @ApiResponse(responseCode = "401", description = "Unauthorized", content = @Content())
+    @ApiResponse(responseCode = "403", description = "Access forbidden", content = @Content())
+    @ApiResponse(responseCode = "412", description = "Reading preconditions failed", content = @Content())
+    @GetMapping("/three-cards-divination/reading")
+    public ThreeCardsDivinationReading performReadingForUser(@Parameter(description = "Basic Auth header")
+                                                             @RequestHeader(value = "Authorization", required = false) String authHeader)
             throws DivinationException, ForbiddenAccessException, UnauthorisedAccessException {
-        String[] credentials = authenticateAndParseHeader(authHeader);
+        String[] credentials = authenticationUtil.validateAndParseHeader(authHeader);
         String username = credentials[0];
         String email = credentials[1];
+        TarotUser tarotUser = new TarotUser(username, email);
         try {
-            return threeCardsDivinationService.getReadingForUser(username, email);
+            return threeCardsDivinationService.getReadingForUser(tarotUser);
         } catch (UserNotFoundException e) {
```

Comment:
**Suggestion:** Add error handling for exceptions thrown in performReadingForUser() method [error handling]
```suggestion
    @GetMapping("/three-cards-divination/reading")
    public ThreeCardsDivinationReading performReadingForUser(@Parameter(description = "Basic Auth header")
                                                             @RequestHeader(value = "Authorization", required = false) String authHeader)
            throws DivinationException, ForbiddenAccessException, UnauthorisedAccessException {
        try {
            String[] credentials = authenticationUtil.validateAndParseHeader(authHeader);
            String username = credentials[0];
            String email = credentials[1];
            TarotUser tarotUser = new TarotUser(username, email);
            return threeCardsDivinationService.getReadingForUser(tarotUser);
        } catch (Exception e) {
            // Handle exception here
            throw new DivinationException("Failed to perform reading");
        }
    }
```